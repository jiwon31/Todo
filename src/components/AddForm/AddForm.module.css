.form {
  display: flex;
  width: 100%;
  padding: 1.4rem 1rem;
  background-color: var(--color-bg);
}

.input {
  flex: 1 0 auto;
  padding: 0.7rem 1rem;
  border: none;
  outline: none;
  font-size: 1.2rem;
  border-top-left-radius: 8px;
  border-bottom-left-radius: 8px;
}

.button {
  padding: 0 2rem;
  font-size: 1.2rem;
  color: var(--color-white);
  background-color: var(--color-accent);
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
}

.button:hover {
  filter: brightness(115%);
}
