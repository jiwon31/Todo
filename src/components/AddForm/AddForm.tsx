import { Todo } from "components/TodoList/TodoList";
import { useState } from "react";
import { v4 as uuidv4 } from "uuid";
import styles from "./AddForm.module.css";

type AddFormProps = {
  onAdd: (todo: Todo) => void;
};

export default function AddForm({ onAdd }: AddFormProps) {
  const [text, setText] = useState("");

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setText(e.target.value);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    // trim() 메서드는 문자열 양 끝의 공백을 제거하고 원본 문자열을 수정하지 않고 새로운 문자열을 반환한다.
    if (text.trim().length === 0) {
      return;
    }
    onAdd({ id: uuidv4(), text, status: "active" });
    setText("");
  };

  return (
    <form className={styles.form} onSubmit={handleSubmit}>
      <input
        className={styles.input}
        type="text"
        placeholder="Add Todo"
        value={text}
        onChange={handleChange}
      ></input>
      <button className={styles.button}>Add</button>
    </form>
  );
}
