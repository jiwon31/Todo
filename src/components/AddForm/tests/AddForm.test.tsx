import AddForm from "components/AddForm/AddForm";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("AddForm", () => {
  it("renders", () => {
    const component = renderer.create(<AddForm onAdd={jest.fn()} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Form Submit", () => {
    let onAdd: () => void;
    let input: HTMLInputElement;
    let button: HTMLButtonElement;

    beforeEach(() => {
      onAdd = jest.fn();
      render(<AddForm onAdd={onAdd} />);
      input = screen.getByPlaceholderText("Add Todo");
      button = screen.getByText("Add");
    });

    it("calls onAdd when button is clicked and valid todo is entered", () => {
      userEvent.type(input, "New Todo");
      userEvent.click(button);

      expect(onAdd).toHaveBeenCalledTimes(1);
    });

    it("does not call onAdd when the todo is empty", () => {
      userEvent.type(input, " ");
      userEvent.click(button);

      expect(onAdd).toHaveBeenCalledTimes(0);
    });
  });
});
