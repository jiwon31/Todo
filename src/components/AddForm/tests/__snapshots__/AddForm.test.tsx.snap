// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddForm renders 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Add Todo"
    type="text"
    value=""
  />
  <button
    className="button"
  >
    Add
  </button>
</form>
`;
